File: semestr-6/Algorytmy i struktury danych 2/lab-4/sorting-network.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

func getCsvData(fileName string) [][]string {
	f, err := os.Open(fileName)
	if err != nil {
		log.Fatal(err)
	}

	defer f.Close()

	csvReader := csv.NewReader(f)
	csvReader.FieldsPerRecord = -1
	data, err := csvReader.ReadAll()
	if err != nil {
		log.Fatal(err)
	}
	return data
}

func getNetwork(filename string) [][][]int {
	data := getCsvData(filename)
	network := make([][][]int, len(data)-2)
	for i, row := range data[2:] {
		network[i] = make([][]int, len(row))
		for j, value := range row {
			value = strings.Trim(value, " ")
			values := strings.Split(value, "-")
			x, _ := strconv.Atoi(values[0])
			y, _ := strconv.Atoi(values[1])
			network[i][j] = make([]int, 2)
			network[i][j][0] = x
			network[i][j][1] = y
		}
	}
	return network
}

func getValues(fname string) []int {
	data := getCsvData(fname)
	values := make([]int, 4)
	elements := strings.Split(data[1][0], " ")
	for i, value := range elements {
		values[i], _ = strconv.Atoi(value)
	}

	return values
}

func processValues(network [][][]int, values []int) []int {
	for _, transform := range network {
		values = processValuesBySingleTransform(transform, values)
	}

	return values
}

func processValuesBySingleTransform(transform [][]int, values []int) []int {
	for _, pair := range transform {
		x, y := pair[0], pair[1]
		if values[x-1] < values[y-1] {
			values[x-1], values[y-1] = values[y-1], values[x-1]
		}
	}

	return values
}

func main() {
	// fmt.Println(network)
	values := getValues("Input_0001.csv")
	network := getNetwork("Network_20_0001.csv")
	// outputData := getCsvData("Output_0001.csv")
	_ = values
	_ = network
	fmt.Println(values)

	values = processValues(network, values)
	fmt.Println(values)
}

File: semestr-6/Algorytmy i struktury danych 2/lab-4/sorting-network.py
import csv
from random import randint

def get_data(filename):
    with open(filename, newline='') as csvfile:
        data = csv.reader(csvfile, delimiter=',')
        data = list([x.strip() for x in row] for row in data)
        csvfile.close()
    return data


def get_transforms_from_network_data(network_data):
    transforms = []
    for row in network_data:
        transform_data = []
        for element in row:
            x, y = element.split('-')
            x = int(x)
            y = int(y)
            transform_data.append((x, y))
        transforms.append(transform_data)
    return transforms

def apply_transform(transform, values):
    for x, y in transform:
        if values[x-1] < values[y-1]:
            values[x-1], values[y-1] = values[y-1], values[x-1]
    return values

def get_network_from_file(fname):
    network_data = get_data(fname)
    transforms = get_transforms_from_network_data(network_data[2:])
    return transforms

def get_values_from_file(fname):
    input_data = get_data(fname)
    values = [int(x) for x in input_data[1:][0][0].split()]
    return values



def process_values(network, values):
    output_data = []
    for transform in network:
        output_data.append(str.join(' ', [f"{i+1}:{values[i]}" for i in range(len(values))]))
        output_data.append(str.join(' ', [f"{x}-{y}" for x,y in transform]))
        values = apply_transform(transform, values)
    output_data.append(str.join(' ', [f"{i+1}:{values[i]}" for i in range(len(values))]))
    
    return values, output_data


network_20 = get_network_from_file("Network_20_0001.csv")
network_42 = get_network_from_file("Network_42_0001.csv")
values = get_values_from_file("Input_0001.csv")

values, output_data = process_values(network_20, values)

values = get_values_from_file("Input_0001.csv")

values = [randint(1,9) for x in range(4)]
values, output_data = process_values(network_42, values)

for row in output_data:
    print(row)
